www-reputation-credentials-protocol.md

```diff
@@ -0,0 +1,265 @@
+	|  Alice  |--------------------------------->|   Bob   |------------------------------------->|  Caroll  |
+	|	  |	                     	     |	       |		   	              |	         |
+	|_________|	                     3b.     |_________|		                      |__________|
+		^				no    /
+		 \	     fail_htlc		     /
+		  \_________________________________/
+			
+```
+
+In this diagram:
+1. Alice sends a HTLC+credentials to Bob.
+2. Bob extracts the credentials, verifies the signature and validate the credentials against his own `routing_policy`.
+3a. The routing checks are successful, Bob forwards the HTLC along the Bob-Caroll link.
+4a. Bob stores the backward credentials into the local state until the HTLC settlement.
+3b. The routing checks are unsuccessful, Bob rejects the HTLC.
+
+The credentials are wrapped inside the HTLC onion `credentials_payload`. There are two types of credentials: forward/backward.
+The forward credentials are covering the ongoing HTLC forward, they should have been counter-signed by the target routing hop
+during the dissemination phase or a previous HTLC forward/settlement pahse with the same target routing hop. The backward
+credentials are also wrapped into the HTLC onion, however they're not signed and should be blinded. If the settlement is successful,
+those backward credentials should be the "forward" credentials for Alice's future HTLC sends.
+
+## Signing or slashing credentials during HTLC Settlement
+
+The current HTLC settlement does not encompass any routing checks processing by routing hops, whatever the outcome `update_fulfill_htlc`/
+`update_fail_htlc`.
+
+```
+						    2. HTLC success ?
+
+							2a. yes, sign backward credentials
+
+							2b. no, slash forward credentials
+
+	 _________   3a. HTLC + onions(signed_credentials)   _________						__________
+	|	  |<----------------------------------------|	      |	       1.			       |	  |
+	|	  |				            |	      |		  HTLC settlement	       |	  |
+	|  Alice  |				            |   Bob   |<---------------------------------------|  Caroll  |
+	|	  |         3b. HTLC 		            |	      |				               |	  |
+	|_________|<----------------------------------------|_________|					       |__________|
+
+```
+
+In this diagram:
+1. Caroll settles the HTLC back to Bob with either a success or a failure.
+2a. If it's a success, Bob sign the blinded backward credentials stored during the previous phase.
+3a. If it's a failure, Bob slashes the forward credentials. There is no backward credentials return back to Alice.
+
+To prevent replay usage of slashed forward credentials or previously used forward credentials, the signature should be
+logged by the routing hop in an [accumulator data structure](https://link.springer.com/content/pdf/10.1007/3-540-48285-7_24.pdf),
+allowing efficient test membership. Merkle tree can constitute such accumulator. To limit the growing size of this accumulator,
```

Comment:
I believe this is a lossy data structure. Does that create issues with false positives or negatives?

Reply:
I think this is correct, if we don't make the merkle branch to the root deterministic on the signature itself. E.g the X first bit of the signature could give you the left-right of merkle path from the root to the leaf. More efficient accumulators, with less studied assumptions, could be used as it can be upgraded in a non-coordinated fashion, this is purely internally to the node.

Reply:
This is indeed probably something that can be solved in the implementation phase. With the expiry, the set cannot grow indefinitely. Is that expiry actually encoded inside the 32 byte credential?

Reply:
See the new gossip in BOLT7 `routing_policy` with the field `credential_expiration_height`. Encoding directly the height in the credentials could be a correlation vector, I believe. Failure to respect its own routing policy by a hop should be likely penalized by routing algorithms.

Reply:
So `credential_expiration_height` is a delta height that both sender and router use to expire tokens? Can the router do this when it gets presented the unblinded token, or how does the router know when the token was produced?

Reply:
Yes, both the sender and the router should consider the credentials as expired once the `credential_expiration_height` is reached. Without assigning signing key per expiration period, I don't think there is a way for the router to dissociate credentials across expiration periods. However, rotating signing key might be undesirable as raised in another comment. Another direction could be to request the senders to "tag" all the credentials starting within said expiration period. For global visibility, the tag should be part of the `routing_policy` (without introducing traffic correlation).  

For now, I think the only clear rational to introduce a `credential_expiration_height` is effectively to bound the anti-replay accumulator size. Though key rotation, upgrading the credential format, or updating core fields of the routing policy could be justification.

Reply:
Can the router still verify the tag after unblinding? I'd think it has become unreadable for the router when it is presented as a forward credential?

Reply:
The HTLC sender generates a credentials, concatenate the current tag for the period, blind and send it to the routing hop for signature. The counter-signed blinded credential is sent back to the HTLC sender, which can unblind both the credential/signature. I think the tag should still bind.